[DeviceMesh] Add get_local_rank() API to DeviceMesh

wz337 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #114710
* __->__ #114709
* #114708

As title.

Differential Revision: [D51625152](https://our.internmc.facebook.com/intern/diff/D51625152/)

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @tianyu-l @wconstab @yf225 @kiukchung @d4l3k @lucasllc

wz337 committed:
[DeviceMesh] Add get_local_rank() API to DeviceMesh

Ad title.

Differential Revision: [D51625152](https://our.internmc.facebook.com/intern/diff/D51625152/)

[ghstack-poisoned]

wanchaol commented on torch/distributed/_device_mesh.py:
```diff
@@ -381,8 +381,35 @@ def shape(self) -> Tuple[int, ...]:
+        self, mesh_dim: Optional[Union[int, str]] = None
+    ) -> int:
+        """
+        Returns the local rank of the given mesh_dim.
```
```suggestion
        Returns the local rank on the given mesh_dim of the DeviceMesh. 
```

wanchaol commented on torch/distributed/_device_mesh.py:
```diff
@@ -381,8 +381,35 @@ def shape(self) -> Tuple[int, ...]:
+            of the mesh dimension. Default is None.
+
+        Returns:
+            An integer denotes the local rank.
```
shall we add a small example in the doc about a 2-D mesh, to describe the meaning of local_rank on mesh dimension?

wanchaol commented on torch/distributed/_device_mesh.py:
```diff
@@ -381,8 +381,35 @@ def shape(self) -> Tuple[int, ...]:
+        """
+        if self.ndim > 1 and mesh_dim is None:
+            raise RuntimeError(
+                f"Found mesh_dim = {self.mesh.ndim}",
```
```suggestion
                f"Found the DeviceMesh have {self.mesh.ndim} dimensions", 
```

wanchaol commented on torch/distributed/_device_mesh.py:
```diff
@@ -381,8 +381,35 @@ def shape(self) -> Tuple[int, ...]:
+        if self.ndim > 1 and mesh_dim is None:
+            raise RuntimeError(
+                f"Found mesh_dim = {self.mesh.ndim}",
+                "Optional kwarg `mesh_dim` needs to be specified when mesh_dim > 1."
```
```suggestion
                "Optional kwarg `mesh_dim` needs to be specified when device_mesh.ndim > 1."
```

wanchaol reviewed (approved):
lgtm, some comment suggestions

wz337 committed:
Update on "[DeviceMesh] Add get_local_rank() API to DeviceMesh"


As title.

Differential Revision: [D51625152](https://our.internmc.facebook.com/intern/diff/D51625152/)

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wconstab mrshenli zhaojuanmao rohan-varma kiukchung d4l3k lucasllc XilunWu tianyu-l

[ghstack-poisoned]

wz337 committed:
Update on "[DeviceMesh] Add get_local_rank() API to DeviceMesh"


As title.

Differential Revision: [D51625152](https://our.internmc.facebook.com/intern/diff/D51625152/)

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wconstab mrshenli zhaojuanmao rohan-varma kiukchung d4l3k lucasllc XilunWu tianyu-l

[ghstack-poisoned]

wz337 committed:
Update on "[DeviceMesh] Add get_local_rank() API to DeviceMesh"


As title.

Differential Revision: [D51625152](https://our.internmc.facebook.com/intern/diff/D51625152/)

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wconstab mrshenli zhaojuanmao rohan-varma kiukchung d4l3k lucasllc XilunWu tianyu-l

[ghstack-poisoned]

fegin commented on torch/distributed/_device_mesh.py:
````diff
@@ -375,8 +375,50 @@ def shape(self) -> Tuple[int, ...]:
+                                [0, 1, 2, 3],
+                                [4, 5, 6, 7]
+                            ])
+            ```
````
hmm, why "```" exist here?

wz337 committed:
Update on "[DeviceMesh] Add get_local_rank() API to DeviceMesh"


As title.

Differential Revision: [D51625152](https://our.internmc.facebook.com/intern/diff/D51625152/)

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wconstab mrshenli zhaojuanmao rohan-varma kiukchung d4l3k lucasllc XilunWu tianyu-l

[ghstack-poisoned]

wz337 committed:
Update on "[DeviceMesh] Add get_local_rank() API to DeviceMesh"


As title.

Differential Revision: [D51625152](https://our.internmc.facebook.com/intern/diff/D51625152/)

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wconstab mrshenli zhaojuanmao rohan-varma kiukchung d4l3k lucasllc XilunWu tianyu-l

[ghstack-poisoned]

wz337 committed:
Update on "[DeviceMesh] Add get_local_rank() API to DeviceMesh"


As title.

Differential Revision: [D51625152](https://our.internmc.facebook.com/intern/diff/D51625152/)

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wconstab mrshenli zhaojuanmao rohan-varma kiukchung d4l3k lucasllc XilunWu tianyu-l

[ghstack-poisoned]

wz337 committed:
Update on "[DeviceMesh] Add get_local_rank() API to DeviceMesh"


As title.

Differential Revision: [D51625152](https://our.internmc.facebook.com/intern/diff/D51625152/)

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wconstab mrshenli zhaojuanmao rohan-varma kiukchung d4l3k lucasllc XilunWu tianyu-l

[ghstack-poisoned]

wz337 committed:
Update on "[DeviceMesh] Add get_local_rank() API to DeviceMesh"


As title.

Differential Revision: [D51625152](https://our.internmc.facebook.com/intern/diff/D51625152/)

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wconstab mrshenli zhaojuanmao rohan-varma kiukchung d4l3k lucasllc XilunWu tianyu-l

[ghstack-poisoned]
